WIP / Place Holder for Blog Post Draft for https://github.com/kubernetes/enhancements/issues/1495 (#28970)
kubernetes/website: content/en/blog/_posts/2021-08-30-volume-populators-alpha.md

---

> title: Volume Populators Redesign
> date: 2021-08-30

Reviewer: Please update the front matter; I suggest:
```suggestion
layout: blog
title: "Kubernetes 1.22: A New Design for Volume Populators"
date: 2021-08-30
slug: volume-populators-redesigned
```
(the title I'm flexible on. I do want to see a `slug` set and the `layout` explicitly declared)

---

**Authors:**
Ben Swartzlander (NetApp)

Kubernetes v1.22 introduces a redesigned approach for volume populators. Originally implemented
in v1.18, the API suffered from backwards compatibility issues and v1.22 includes a new API
field called `dataSourceRef` that fixes these problems.

## Data sources

> The PersistentVolumeClaim object got a new field called `dataSource` when snapshot and cloning
> support was added to Kubernetes. This field allowed new PVCs to point to other PVCs or
> VolumeSnapshots in the same namespace, and modified the normal provisioning process so that

Reviewer: ```suggestion
Earlier Kubernetes releases already added a `dataSource` field into the the
[PersistentVolumeClaim](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims) API, used for snapshots and for cloning quiescent volumes.
You could use the `dataSource` field when creating a new PVC, referencing either
an existing PVC or a VolumeSnapshot in the same namespace. That also
modified the normal provisioning process so that
```

instead of yielding an empty volume, the new PVC contained the same data as either the cloned
PVC or the cloned VolumeSnapshot.

> Volume populators embrace the same design idea, but extend it to any type of object, as long
> as there exists a CRD to define the data source and a populator controller to implement the
> logic. Initially the `dataSource` field was directly extended to allow arbitrary objects, if
> the `AnyVolumeDataSource` feature gate was enabled on the cluster. That change unfortunately
> caused backwards compatibility problems, and so the new `dataSourceRef` field was born.

Reviewer: ```suggestion
Volume populators embrace the same design idea, but extend it to any type of object, as long
as there exists a [custom resource](/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
to define the data source, and a populator controller to implement the logic logic. Initially,
the `dataSource` field was directly extended to allow arbitrary objects, if the `AnyVolumeDataSource`
feature gate was enabled on a cluster. That change unfortunately caused backwards compatibility
problems, and so the new `dataSourceRef` field was born.
```

In v1.22 if the `AnyVolumeDataSource` feature gate is enabled, the `dataSourceRef` field is
added, which behaves similarly to the `dataSource` field except that it allows arbitrary
objects to be specified. The API server ensures that the two fields always have the same
contents, and neither of them are mutable. The differences is that at creation time
`dataSource` allows only PVCs or VolumeSnapshots, and ignores all other values, while
`dataSourceRef` allows most types of objects, and in the few cases it doesn't allow an
object (core objects other than PVCs) a validation error occurs.

This API redesign will eventually result in a preference for the `dataSourceRef` field for

> all use cases, but in the immediate timeframe it is used by volume populators.

Reviewer 2: I'm not sure what you are trying to say here.  You meant dataSourceRef is used only for create volume from AnyVolumeDataSource now, but eventually dataSourceRef will be the preferred one for create PVC from VolumeSnapshot, cloning, and volume populators?

Reviewer 3: What I'm trying to say is that for clusters where the feature gate is enabled, dataSourceRef is the preferred API. In the future (wen the feature is GA) that will be all clusters. However that's the vast minority of clusters while the feature is Alpha.

> This API redesign will eventually result in a preference for the `dataSourceRef` field for
> all use cases, but in the immediate timeframe it is used by volume populators.

Reviewer: How about:
```suggestion
Within SIG Storage, our plan is for this API change to graduate to stable, meaning that we would
deprecate using `dataSource` and recommend using `dataSourceRef` field for all use cases.
In the v1.22 release, `dataSourceRef` is available (as an alpha feature) specifically for cases
where you want to use for custom volume populators.
```

Reviewer 3: I love this wording, thank you!

## Using populators

Every volume populator must have one or more CRDs that it supports. Administrators may
install the CRD and the populator controller and then PVCs with a `dataSourceRef` specifies
a CR of the type that the populator supports will be handled by the populator controller

> instead of the CSI plugin directly.

Reviewer 2: I think we usually call it CSI driver.

> Underneath the covers, the CSI plugin is still invoked to create an empty volume, which

Reviewer 2: CSI plugin -> CSI driver

the populator controller fills with the appropriate data. The PVC doesn't bind to the PV
until it's fully populated, so it's safe to define a whole application manifest including

> pod and PVC specs and the pods won't being running until everything is ready, just as if

Reviewer 2: being running -> be running

the PVC was a clone of another PVC or VolumeSnapshot.

## How it works

PVCs with data sources are still noticed by the external-provisioner sidecar for the
related storage class (assuming a CSI provisioner is used), but because the sidecar
doesn't understand the data source kind, it doesn't do anything. The populator controller
is also watching for PVCs with data sources of a kind that it understands and when it
sees one, it creates a temporary volume of the same size, volume mode, storage class,
and even on the same topology (if topology is used) as the original volume. The populator
controller creates a worker pod that attaches to the volume and writes the necessary
data to it, then detaches from the volume and the populator controller rebinds the
temporary volume to the orignal PVC.

> ## Important components

Reviewer: ```suggestion
## Trying it out
```
(I _hope_ folks aren't going to jump into a production deploy whilst reading this&hellip;)

> The following things are required to used volume populators:

Reviewer 2: to used -> to use

* Enable the `AnyVolumeDataSource` feature gate
* Install a CRD for the specific data source / populator
* Install the populator controller itself

Populator controllers may use the [lib-volume-populator](https://github.com/kubernetes-csi/lib-volume-populator)
library to do most of the Kubernetes API level work. Individual populators only need to
provide logic for actually writing data into the volume based on a particular CR
instance. This library provides a sample populator implementation.

> These optional components improve user feedback:

Reviewer 2: improve user feedback -> improve user experience?

Reviewer 2: Can you address this comment?

* Install the VolumePopulator CRD

> * Create an VolumePopulator CR for the specific data source

Reviewer: ```suggestion
* Create a VolumePopulator custom resource for your specific data source
```

* Install the [volume data source validator](https://github.com/kubernetes-csi/volume-data-source-validator)
  controller (alpha)

The purpose of these components is to generate warning events on PVCs with data sources
for which there is no populator.

## Putting it all together

To see how this works, you can install the sample "hello" populator and try it
out.

> First we will install the volume-data-source-validator controller.

Reviewer 2: First we will install -> First install

```terminal
$ kubectl apply -f https://github.com/kubernetes-csi/volume-data-source-validator/blob/master/deploy/kubernetes/rbac-data-source-validator.yaml

> $ kubectl apply -f https://github.com/kubernetes-csi/volume-data-source-validator/blob/master/deploy/kubernetes/setup-data-source-validator.yaml

Reviewer: I'd omit the prompt here: it makes copying and pasting easier.

Also, consider using a git ref other than `master` (we don't keep articles up to date the way we maintain docs, so a point-in-time snapshot feels appropriate).

Reviewer 3: For this, we need to release the updated repos to get a tag to refer to. That work is happening in parallel. I agree that the master branch is not ideal.

```

> Next we will install the example populator.

Reviewer 2: Next we will install -> Next install

```terminal
$ kubectl apply -f https://github.com/kubernetes-csi/lib-volume-populator/blob/master/example/hello-populator/crd.yaml
$ kubectl apply -f https://github.com/kubernetes-csi/lib-volume-populator/blob/master/example/hello-populator/deploy.yaml
```

> We will create an instance of the `Hello` CR, with some text.

Reviewer 2: We will create -> Create

```yaml
apiVersion: hello.k8s.io/v1alpha1
kind: Hello
metadata:
  name: example-hello
spec:
  fileName: example.txt
  fileContents: Hello, world!
```

> We will create a PVC that refers to that CR as its data source.

Reviewer 2: We will create -> Create

Reviewer: or “The following manifest creates a PVC&hellip;”

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: example-pvc
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 10Mi
  dataSourceRef:
    apiGroup: hello.k8s.io
    kind: Hello
    name: example-hello
  volumeMode: Filesystem
```

> We will run a job that reads the file in the PVC.

Reviewer 2: We will run -> Run

Reviewer: I'd write:
```suggestion
Next, run a Job that reads the file in the PVC:
```

          volumeMounts:
            - name: vol
              mountPath: /mnt
      restartPolicy: Never
      volumes:
        - name: vol
          persistentVolumeClaim:
            claimName: example-pvc
```

Wait for the job to complete (including all of its dependencies).

```terminal
$ kubectl wait --for=condition=Complete job/example-job
```

And last examine the log from the job.

```terminal
$ kubectl logs job/example-job
Hello, world!
```

Note that the volume already contained a text file with the string contents from
the CR. This is only the simplest example. Actual populators can set up the volume
to contain arbitrary contents.

## How to write your own volume populator

Developers interested in writing new poplators are encouraged to use the
[lib-volume-populator](https://github.com/kubernetes-csi/lib-volume-populator) library
and to only supply a small controller wrapper around the library, and a pod image
capable of attaching to volumes and writing the appropriate data to the volume.

Individual populators can be extremely generic such that they work with every type
of PVC, or they can do vendor specific things to rapidly fill a volume with data

> of the volume was provisioned by a specific CSI driver from the same vendor, for

Reviewer 2: of the volume -> if the volume?

example, by communicating directly with the storage for that volume.

## The future

As this feature is still in alpha, we expect to update the out of tree controllers
with more tests and documentation. The community plans to eventually re-implement
the populator library as a sidecar, for ease of operations.

We hope to see some official community-supported populators for some widely-shared
use cases. Also, we expect that volume populators will be used by backup vendors
as a way to "restore" backups to volumes, and possibly a standardized API to do
this will evolve.

## How can I learn more?

> Please read the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators).

Reviewer: Could also link to https://kubernetes.io/docs/concepts/storage/persistent-volumes/#volume-populators-and-data-sources

---
layout: blog
title: "Kubernetes 1.22: A New Design for Volume Populators"
date: 2021-08-30
slug: volume-populators-redesigned
---

**Authors:**
Ben Swartzlander (NetApp)

> Kubernetes v1.22 introduces a redesigned approach for volume populators. Originally implemented

Reviewer: ```suggestion
Kubernetes v1.22, released earlier this month, introduced a redesigned approach
for volume populators. Originally implemented
```
?

> in v1.18, the API suffered from backwards compatibility issues and v1.22 includes a new API

Reviewer: ```suggestion
in v1.18, the API suffered from backwards compatibility issues. Kubernetes v1.22 includes a new API
```

field called `dataSourceRef` that fixes these problems.

## Data sources

Earlier Kubernetes releases already added a `dataSource` field into the the
[PersistentVolumeClaim](/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims) API,

> used for snapshots and for cloning quiescent volumes. You could use the `dataSource` field when

Reviewer 3: Good catch, this text was copy/pasted and I missed that word.

as there exists a [custom resource](/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
to define the data source, and a populator controller to implement the logic logic. Initially,
the `dataSource` field was directly extended to allow arbitrary objects, if the `AnyVolumeDataSource`
feature gate was enabled on a cluster. That change unfortunately caused backwards compatibility
problems, and so the new `dataSourceRef` field was born.

In v1.22 if the `AnyVolumeDataSource` feature gate is enabled, the `dataSourceRef` field is
added, which behaves similarly to the `dataSource` field except that it allows arbitrary
objects to be specified. The API server ensures that the two fields always have the same
contents, and neither of them are mutable. The differences is that at creation time
`dataSource` allows only PVCs or VolumeSnapshots, and ignores all other values, while
`dataSourceRef` allows most types of objects, and in the few cases it doesn't allow an
object (core objects other than PVCs) a validation error occurs.

Within SIG Storage, our plan is for this API change to graduate to stable, meaning that we would

> deprecate using `dataSource` and recommend using `dataSourceRef` field for all use cases.

Reviewer 2: Can you clarify this as follows:
```
When this API change graduates to stable, we would deprecate using `dataSource` and recommend using `dataSourceRef` field for all use cases.
```

where you want to use for custom volume populators.

## Using populators

Every volume populator must have one or more CRDs that it supports. Administrators may
install the CRD and the populator controller and then PVCs with a `dataSourceRef` specifies
a CR of the type that the populator supports will be handled by the populator controller
instead of the CSI driver directly.

Underneath the covers, the CSI driver is still invoked to create an empty volume, which
the populator controller fills with the appropriate data. The PVC doesn't bind to the PV
until it's fully populated, so it's safe to define a whole application manifest including
pod and PVC specs and the pods won't begin running until everything is ready, just as if
the PVC was a clone of another PVC or VolumeSnapshot.

## How it works

PVCs with data sources are still noticed by the external-provisioner sidecar for the
related storage class (assuming a CSI provisioner is used), but because the sidecar
doesn't understand the data source kind, it doesn't do anything. The populator controller
is also watching for PVCs with data sources of a kind that it understands and when it

> sees one, it creates a temporary volume of the same size, volume mode, storage class,

Reviewer 3: The temporary object is actually a PVC -- the resulting PV and physical volume aren't temporary because they later get bound to the original PVC. I will clean up this language.

Developers interested in writing new poplators are encouraged to use the
[lib-volume-populator](https://github.com/kubernetes-csi/lib-volume-populator) library
and to only supply a small controller wrapper around the library, and a pod image
capable of attaching to volumes and writing the appropriate data to the volume.

Individual populators can be extremely generic such that they work with every type
of PVC, or they can do vendor specific things to rapidly fill a volume with data
if the volume was provisioned by a specific CSI driver from the same vendor, for
example, by communicating directly with the storage for that volume.

## The future

As this feature is still in alpha, we expect to update the out of tree controllers
with more tests and documentation. The community plans to eventually re-implement
the populator library as a sidecar, for ease of operations.

We hope to see some official community-supported populators for some widely-shared
use cases. Also, we expect that volume populators will be used by backup vendors
as a way to "restore" backups to volumes, and possibly a standardized API to do
this will evolve.

## How can I learn more?

> Please read the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators).

Reviewer: ```suggestion
The enhancement proposal,
[Volume Populators](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators), includes lots of detail about the history and technical implementation
of this feature.
```

> User docs are available [here](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#volume-populators-and-data-sources)

Reviewer: ```suggestion
[Volume populators and data sources] (in the documenation topic about persistent volumes)
explains how to use this feature in your cluster.
```

used for cloning volumes and creating volumes from snapshots. You could use the `dataSource` field when
creating a new PVC, referencing either an existing PVC or a VolumeSnapshot in the same namespace.
That also modified the normal provisioning process so that instead of yielding an empty volume, the
new PVC contained the same data as either the cloned PVC or the cloned VolumeSnapshot.

Volume populators embrace the same design idea, but extend it to any type of object, as long
as there exists a [custom resource](/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
to define the data source, and a populator controller to implement the logic. Initially,
the `dataSource` field was directly extended to allow arbitrary objects, if the `AnyVolumeDataSource`
feature gate was enabled on a cluster. That change unfortunately caused backwards compatibility
problems, and so the new `dataSourceRef` field was born.

In v1.22 if the `AnyVolumeDataSource` feature gate is enabled, the `dataSourceRef` field is
added, which behaves similarly to the `dataSource` field except that it allows arbitrary
objects to be specified. The API server ensures that the two fields always have the same

> contents, and neither of them are mutable. The differences is that at creation time

Reviewer 2: The differences is -> The differences are ?

capable of attaching to volumes and writing the appropriate data to the volume.

Individual populators can be extremely generic such that they work with every type
of PVC, or they can do vendor specific things to rapidly fill a volume with data
if the volume was provisioned by a specific CSI driver from the same vendor, for
example, by communicating directly with the storage for that volume.

## The future

As this feature is still in alpha, we expect to update the out of tree controllers
with more tests and documentation. The community plans to eventually re-implement
the populator library as a sidecar, for ease of operations.

We hope to see some official community-supported populators for some widely-shared
use cases. Also, we expect that volume populators will be used by backup vendors
as a way to "restore" backups to volumes, and possibly a standardized API to do
this will evolve.

## How can I learn more?

The enhancement proposal,
[Volume Populators](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators), includes lots of detail about the history and technical implementation
of this feature.

> [Volume populators and data sources] (in the documenation topic about persistent volumes)

Reviewer: I noticed a broken hyperlink
```suggestion
[Volume populators and data sources](/docs/concepts/storage/persistent-volumes/#volume-populators-and-data-sources), within the documentation topic about persistent volumes,
```
